File: integration/testdata/go/database/echo.go
//ftl:module echo
package echo

import (
	"context"

	"github.com/TBD54566975/ftl/go-runtime/ftl" // Import the FTL SDK.
)

var db = ftl.PostgresDatabase("testdb")

type InsertRequest struct {
	Data string
}

type InsertResponse struct{}

//ftl:verb
func Insert(ctx context.Context, req InsertRequest) (InsertResponse, error) {
	err := persistRequest(req)
	if err != nil {
		return InsertResponse{}, err
	}

	return InsertResponse{}, nil
}

func persistRequest(req InsertRequest) error {
	_, err := db.Exec(`CREATE TABLE IF NOT EXISTS requests
	       (
	         data TEXT,
	         created_at TIMESTAMPTZ NOT NULL DEFAULT (NOW() AT TIME ZONE 'utc'),
	         updated_at TIMESTAMPTZ NOT NULL DEFAULT (NOW() AT TIME ZONE 'utc')
	      );`)
	if err != nil {
		return err
	}
	_, err = db.Exec("INSERT INTO requests (data) VALUES ($1);", req.Data)
	if err != nil {
		return err
	}
	return nil
}

File: integration/testdata/kotlin/database/Echo.kt
package ftl.echo

import ftl.builtin.Empty
import xyz.block.ftl.Context
import xyz.block.ftl.Verb
import xyz.block.ftl.Database

data class InsertRequest(val data: String)

val db = Database("testdb")

@Verb
fun insert(context: Context, req: InsertRequest): Empty {
  persistRequest(req)
  return Empty()
}

fun persistRequest(req: InsertRequest) {
  db.conn {
    it.prepareStatement(
      """
        CREATE TABLE IF NOT EXISTS requests
        (
          data TEXT,
          created_at TIMESTAMPTZ NOT NULL DEFAULT (NOW() AT TIME ZONE 'utc'),
          updated_at TIMESTAMPTZ NOT NULL DEFAULT (NOW() AT TIME ZONE 'utc')
       );
       """
    ).execute()
    it.prepareStatement("INSERT INTO requests (data) VALUES ('${req.data}');")
      .execute()
  }
}
